### PR TITLE
Use context-dependent BoundManager to store bounds in Tableau

### PR DESCRIPTION
**Pull request by AleksandarZeljic**
This PR transitions Marabou to the first stage of using Context to synchronize the search. Context is added to the SmtCore which controls the synchronization, enabling the Tableau to use context dependent BoundManager to store the bounds and removing the memory management overhead. 

The PR includes the following changes:
- Introduce IBoundManager and MockBoundManager.
- Adds local bounds to BoundManager for expediency of access by Tableau.
- Use BoundManager to store bounds in Tableau.
- Rename and move Tableau::_validBounds to BoundManager::_consistentBounds.
- BoundManager::set*Bounds methods no longer throw InfeasibleQueryException and set the _consistentBounds flag instead.
- Add Context to SmtCore and synchronize with the search correctly using postContextHook to maintain Tableau invariants.
-  SmtCore continues backtracking immediately if implied CaseSplit results in inconsistent bounds.
- Remove local bounds in the Tableau.
- Initialize BoundManager correctly.

### CODE REVIEW

File: src/engine/BoundManager.cpp [hunk 1]
```diff
@@ -27,7 +27,10 @@ BoundManager::BoundManager( Context &context )
     : _context( context )
     , _size( 0 )
     , _tableau( nullptr )
+    , _consistentBounds( &_context )
+    , _firstInconsistentTightening( 0, 0.0, Tightening::LB )
 {
+  _consistentBounds = true;
```
Comment (anwu1219) [inline src/engine/BoundManager.cpp:33]: Use 4-space instead for 2-space for indentation to be consistent with rest of the code base?

File: src/engine/BoundManager.cpp [hunk 2]
```diff
@@ -94,6 +97,15 @@ bool BoundManager::tightenUpperBound( unsigned variable, double value )
     return tightened;
 }
 
+void BoundManager::recordInconsistentBounds( unsigned variable, double value, Tightening::BoundType type )
```
Comment (anwu1219) [inline src/engine/BoundManager.cpp:100]: Bounds=> Bound ? 
Since we are only recording one bound here.

File: src/engine/Engine.h [hunk 3]
```diff
@@ -719,6 +719,13 @@ class Engine : public IEngine, public SignalHandler::Signalable
     */
     bool minimizeCostWithGurobi( const LinearExpression &costFunction );
 
+    /*
+      Get Context reference
+    */
+    Context &getContext() { return _context; }
+
+    bool consistentBounds() const;
```
Comment (anwu1219) [inline src/engine/Engine.h:727]: This method doesn't seem to be used anywhere.

File: src/engine/Tableau.h [hunk 4]
```diff
@@ -118,7 +113,6 @@ class Tableau : public ITableau, public IBasisFactorization::BasisColumnOracle
     unsigned getN() const;
 
     /*
-      Check if an assignment exists for the variable.
```
Comment (anwu1219) [inline src/engine/Tableau.h:121]: Was this line intentionally removed?

File: src/engine/BoundManager.cpp (hunk 1 above)
Comment (AleksandarZeljic) [reply src/engine/BoundManager.cpp:33]: Done.

File: src/engine/Engine.h (hunk 3 above)
Comment (AleksandarZeljic) [reply src/engine/Engine.h:727]: Both methods are used in the SmtCore, getContext() in the ctor and consistentBounds() on line 287.

File: src/engine/Tableau.h (hunk 4 above)
Comment (AleksandarZeljic) [reply src/engine/Tableau.h:121]: Oops, n - restored.

**Review by anwu1219 — APPROVED** [review]
LGTM

**Pull request merged**